<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Tailor by jhhung</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js"></script>
    <script src="javascripts/main.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>
<!--   Code for google analytics -->
  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-53524033-4', 'auto');
  ga('send', 'pageview');

</script>
  <body>

      <header>
        <h1>Tailor</h1>
        <p>Tailor is a fast short read aligner like bowtie/BWA/SOAP2 but is designed specifically to discovering trimming and tailing events for small silencing RNAs.</p>
      </header>

      <div id="banner">
        <span id="logo"></span>

        <a href="https://github.com/jhhung/Tailor" class="button fork"><strong>View On GitHub</strong></a>
        <div class="downloads">
          <span>Downloads:</span>
          <ul>
            <li><a href="https://github.com/jhhung/Tailor/zipball/master" class="button">ZIP</a></li>
            <li><a href="https://github.com/jhhung/Tailor/tarball/master" class="button">TAR</a></li>
          </ul>
        </div>
      </div><!-- end banner -->

    <div class="wrapper">
      <nav>
        <ul></ul>
      </nav>
      <section>
        <h1>
<a name="tailor" class="anchor" href="#tailor"><span class="octicon octicon-link"></span></a>Tailor</h1>

<p>Tailor is a Burrowsâ€“Wheeler transform (BWT) based fast aligner specialized in discovering non-templated addition of nucleotide to the 3' end of nucleic acids (a.k.a., tailing) from Next Generation Sequencing data.  </p>

<p>Tailor is released under GPLv2 with additional restriction so that is only applicable to individuals and non-profits and that any for-profit company must purchase a different license.   </p>

<p>A Shell based pipelines are provided using fastq as input and produce publication quality figures.    </p>

<h2>
<a name="install" class="anchor" href="#install"><span class="octicon octicon-link"></span></a>INSTALL</h2>

<p>*Only 64 bits systems are able to compile and run Tailor.</p>

<h3>
<a name="run-the-binary-directly-without-installation" class="anchor" href="#run-the-binary-directly-without-installation"><span class="octicon octicon-link"></span></a>Run the binary directly without installation</h3>

<p>Please try the precompiled binaries first, most of the linux systems should be able to run Tailor without any troubles.</p>

<div class="highlight highlight-bash"><pre>bin/tailor_linux        <span class="c"># for linux</span>
bin/tailor_mac_x64      <span class="c"># for OSX</span>
</pre></div>

<p>Or you can find them in the release tab in this page or at this <a href="https://github.com/jhhung/Tailor/releases">link</a>.</p>

<h3>
<a name="compile-from-the-source-code" class="anchor" href="#compile-from-the-source-code"><span class="octicon octicon-link"></span></a>Compile from the source code</h3>

<h4>
<a name="install-the-dependencies" class="anchor" href="#install-the-dependencies"><span class="octicon octicon-link"></span></a>Install the dependencies</h4>

<ul>
<li>1.1 Relative recent C++ compiler that support most features of C++11. We recommend <a href="http://gcc.gnu.org/">GCC</a>.</li>
<li>1.2 <a href="http://www.boost.org/users/download/">Boost</a>
</li>
<li>1.3 <a href="http://www.cmake.org/">CMake</a>
</li>
</ul><h4>
<a name="get-the-latest-version-of-the-software" class="anchor" href="#get-the-latest-version-of-the-software"><span class="octicon octicon-link"></span></a>Get the latest version of the software</h4>

<div class="highlight highlight-bash"><pre>git clone git@github.com:jhhung/Tailor.git
</pre></div>

<h4>
<a name="enter-the-folder-tailor-and" class="anchor" href="#enter-the-folder-tailor-and"><span class="octicon octicon-link"></span></a>Enter the folder Tailor and:</h4>

<ul>
<li>Set environmental variable <code>$BOOST_ROOT</code> to the directory of boost if CMake cannot find boost automatically;</li>
<li>Set environmental variable <code>$CC</code> and <code>$CXX</code> to the gcc/g++ compiler you want to use.<br>
</li>
</ul><div class="highlight highlight-bash"><pre>cmake .
</pre></div>

<h4>
<a name="compile-the-software-by-typing" class="anchor" href="#compile-the-software-by-typing"><span class="octicon octicon-link"></span></a>Compile the software by typing:</h4>

<div class="highlight highlight-bash"><pre>make
</pre></div>

<h4>
<a name="troubleshooting" class="anchor" href="#troubleshooting"><span class="octicon octicon-link"></span></a>troubleshooting</h4>

<ul>
<li>If you got linker error, it is possible that the default library in the lib/ is not suitable to your platform. 
There are two libraries available, one is for Mac OSX one is for Linux, rename the one that fit the best to "libabwt_table.a",
and retype </li>
</ul><div class="highlight highlight-bash"><pre>make
</pre></div>

<h2>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>USAGE</h2>

<hr><h3>
<a name="tailor-1" class="anchor" href="#tailor-1"><span class="octicon octicon-link"></span></a>tailor</h3>

<h4>
<a name="build-genomic-index-similar-to-bowtie-build" class="anchor" href="#build-genomic-index-similar-to-bowtie-build"><span class="octicon octicon-link"></span></a>Build genomic index (similar to bowtie-build)</h4>

<div class="highlight highlight-bash"><pre>tailor build -i genome.fa -p genome
</pre></div>

<h4>
<a name="mapping" class="anchor" href="#mapping"><span class="octicon octicon-link"></span></a>Mapping</h4>

<div class="highlight highlight-bash"><pre>tailor map -p genome -n 8 -i smallRNA.fq
</pre></div>

<hr><h3>
<a name="tailing-pipeline" class="anchor" href="#tailing-pipeline"><span class="octicon octicon-link"></span></a>tailing pipeline</h3>

<div class="highlight highlight-bash"><pre><span class="c"># reads.fq: input reads in fastq format: </span>
<span class="c"># dm3.fa: genome sequence in fasta format; it will generate index in the index folder of tailor directory, if it doesn't exist</span>
<span class="c"># genomic_feature_file; used to generate figures for different genomic features (exon, intron...). See our example in the annotation folder to make such file</span>
<span class="c"># using 24 CPUs</span>
<span class="c"># use PHRED score 20 as filter: only reads with every base equal to to higher than 20 pass the filter and enters the pipeline</span>
run_miRNA_tailing_pipeline.sh <span class="se">\ </span>
    -i reads.fq  <span class="se">\ </span>
    -g dm3.fa <span class="se">\ </span>
    -t genomic_feature_file <span class="se">\ </span>
    -o output_dir <span class="se">\ </span>
    -c 24 <span class="se">\ </span>
    -q 20
</pre></div>

<h2>
<a name="download" class="anchor" href="#download"><span class="octicon octicon-link"></span></a>Download</h2>

<hr><h4>
<a name="indexes" class="anchor" href="#indexes"><span class="octicon octicon-link"></span></a>Indexes</h4>

<ul>
<li>Tailor indexes</li>
</ul><div class="highlight highlight-bash"><pre><span class="c"># You can find all the pre-bulit indexes in:</span>
http://www.jhhlab.tw/Tailor/index/

<span class="c"># Human:</span>
http://www.jhhlab.tw/Tailor/index/hg18.tar.gz
http://www.jhhlab.tw/Tailor/index/hg19.tar.gz

<span class="c"># Mouse:</span>
http://www.jhhlab.tw/Tailor/index/mm9.tar.gz
http://www.jhhlab.tw/Tailor/index/mm10.tar.gz

<span class="c"># for downloading</span>
lftp -c <span class="s2">"pget -n 4 http://www.jhhlab.tw/Tailor/index/hg18.tar.gz"</span>
</pre></div>

<hr><h4>
<a name="speed-test-files-for-the-publication" class="anchor" href="#speed-test-files-for-the-publication"><span class="octicon octicon-link"></span></a>Speed test files for the publication</h4>

<ul>
<li>You can download all the related files for the speed test from the link</li>
</ul><pre><code>http://www.jhhlab.tw/Tailor/speed_test_samples/
</code></pre>

<ul>
<li>And the links of original data for non-tailed and tailed reads</li>
</ul><pre><code>http://www.jhhlab.tw/Tailor/speed_test_samples/Drosophila_melanogaster.2m.fq
http://www.jhhlab.tw/Tailor/speed_test_samples/Drosophila_melanogaster.all.randomeTailed.fq
</code></pre>

<ul>
<li>And the speed test log (3 times)</li>
</ul><pre><code>http://www.jhhlab.tw/Tailor/speed_test_samples/test_speed.log
http://www.jhhlab.tw/Tailor/speed_test_samples/test_speed.log2
http://www.jhhlab.tw/Tailor/speed_test_samples/test_speed.log3
</code></pre>

<ul>
<li>And the speed log for bowtie tailing</li>
</ul><pre><code>http://www.jhhlab.tw/Tailor/speed_test_samples/tailing.log
http://www.jhhlab.tw/Tailor/speed_test_samples/tailing.log2
http://www.jhhlab.tw/Tailor/speed_test_samples/tailing.log3
</code></pre>

<ul>
<li>All scripts for speed test can be found in the <code>utils</code> directory</li>
</ul><h2>
<a name="citing-tailor" class="anchor" href="#citing-tailor"><span class="octicon octicon-link"></span></a>Citing Tailor</h2>

<ul>
<li>in preparation</li>
</ul><h2>
<a name="contact" class="anchor" href="#contact"><span class="octicon octicon-link"></span></a>Contact</h2>

<div class="highlight highlight-bash"><pre>    Jui-Hung Hung &lt;juihunghung <span class="sb">`</span>at<span class="sb">`</span> gmail.com&gt;
    Chou Min-Te &lt;poi5305 <span class="sb">`</span>at<span class="sb">`</span> gmail.com&gt;
    Bo W Han &lt;bowhan <span class="sb">`</span>at<span class="sb">`</span> me.com&gt;
</pre></div>
      </section>
      <footer>
        <p>Project maintained by <a href="https://github.com/jhhung">jhhung</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://twitter.com/michigangraham">mattgraham</a></small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
    
  </body>
</html>